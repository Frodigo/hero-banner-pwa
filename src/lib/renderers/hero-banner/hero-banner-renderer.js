import React from 'react';
import HeroBannerFactory from "./factory";
import { contentTypes }from './config';

function HeroBannerRenderer(props) {
    const contentTypeStructureObj = {
        children: []
    };

    const container = new DOMParser().parseFromString(props.html, 'text/html');
    const content =  walk(container.body, contentTypeStructureObj);

    return content.children.map((child, i) => {
        return (
            <div key={i}>
                <HeroBannerFactory data={child} />
            </div>
        );
    });
}

export const Component = HeroBannerRenderer;

export function canRender(content) {
    return /class="hero-block"/gim.test(content);
}

const createContentObject = (type, node) => {
    return {
        contentType: type,
        html: node,
        children: []
    };
};

export const walk = (rootEl, contentTypeStructureObj) => {
    const tree = document.createTreeWalker(
        rootEl,
        NodeFilter.SHOW_ELEMENT | NodeFilter.SHOW_TEXT
    );

    let currentNode = tree.nextNode();
    while (currentNode) {
        if (currentNode.nodeType !== Node.ELEMENT_NODE) {
            currentNode = tree.nextNode();
            continue;
        }

        const currentNodeClassList = currentNode.classList;

        if (!contentTypes.some((item) => {
            return currentNodeClassList.contains((item.cssClass));
        })) {
            currentNode = tree.nextNode();
            continue;
        }

        const props = createContentObject(currentNodeClassList.value, currentNode);
        contentTypeStructureObj.children.push(props);
        walk(currentNode, props);
        currentNode = tree.nextSibling();
    }

    return contentTypeStructureObj;
};
