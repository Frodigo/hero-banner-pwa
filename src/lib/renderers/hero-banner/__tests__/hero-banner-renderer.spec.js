import { walk } from './../hero-banner-renderer'
import renderer from 'react-test-renderer';
import { canRender, Component as HeroBannerRenderer } from '../hero-banner-renderer'
import React from 'react';

jest.mock('@magento/venia-ui/lib/classify');
const mockedHtml = '<div class="hero-block"></div>';

const contentTypeStructureObj = {
    children: []
};

test('Should render if there is <div class="hero-block"></div> in HTML', () => {
    const shouldRender = canRender(mockedHtml);
    expect(shouldRender).toBeTruthy();
});

test('Has hero-banner as a root element', () => {
    const container = new DOMParser().parseFromString(mockedHtml, 'text/html');
    const content  = walk(container, contentTypeStructureObj);

    expect(content.children[0].contentType).toEqual('hero-block');
});

test('Parse valid HeroBanner element', () => {
    const tree = renderer
        .create(<HeroBannerRenderer html={mockedHtml}/>)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('No render if html is broken', () => {
    const tree = renderer
        .create(<HeroBannerRenderer html={'broken'}/>)
        .toJSON();

    expect(tree).toBeNull();
});
