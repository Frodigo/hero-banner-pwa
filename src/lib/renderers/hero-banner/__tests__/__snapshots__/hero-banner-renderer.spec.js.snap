// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse valid HeroBanner element 1`] = `
<div>
  <div
    className="heroBlock"
  />
</div>
`;
